src/clients/UBAClient/UBAClientAbstract.ts

```diff
@@ -0,0 +1,236 @@
+import winston from "winston";
+import { RefundRequestWithBlock, UbaFlow } from "../../interfaces";
+import { BigNumber } from "ethers";
+import { UBAActionType, UBAFeeSpokeCalculator } from "../../UBAFeeCalculator";
+import { RelayerFeeDetails } from "../../relayFeeCalculator";
+import { toBN } from "../../utils";
+
+export type RequestValidReturnType = { valid: false; reason: string } | { valid: true };
+export type OpeningBalanceReturnType = { blockNumber: number; spokePoolBalance: BigNumber };
+export type BalancingFeeReturnType = { balancingFee: BigNumber; actionType: UBAActionType };
+export type SystemFeeResult = { lpFee: BigNumber; depositBalancingFee: BigNumber; systemFee: BigNumber };
+export type RelayerFeeResult = {
+  relayerGasFee: BigNumber;
+  relayerCapitalFee: BigNumber;
+  relayerBalancingFee: BigNumber;
+  relayerFee: BigNumber;
+};
+
+/**
+ * UBAClient is a base class for UBA functionality. It provides a common interface for UBA functionality to be implemented on top of or extended.
+ * This class is not intended to be used directly, but rather extended by other classes that implement the abstract methods.
+ */
+export abstract class BaseUBAClient {
+  protected spokeUBAFeeCalculators: { [chainId: number]: { [token: string]: UBAFeeSpokeCalculator } };
+
+  protected constructor(protected readonly chainIdIndices: number[], protected readonly logger?: winston.Logger) {
+    this.spokeUBAFeeCalculators = {};
+  }
+
+  /**
+   * Retrieves the opening balance for a given token on a given chainId at a given block number
+   * @param chainId The chainId to get the opening balance for
+   * @param spokePoolToken The token to get the opening balance for
```

Comment:
It'd be helpful to clarify that this is the token address, rather than the symbol :+1: